rfcs/0059-dml-atomicity.md

```diff
@@ -0,0 +1,47 @@
+---
+feature: dml_atomicity
+authors:
+  - "DylanChen"
+start_date: "2023/04/10"
+---
+
+# DML Atomicity
+
+## Motivation
+
+As we all know, DML in RisingWave currently cannot provide atomicity. For example, when using the INSERT SELECT statement to load a large amount of data from one table to another, it is possible for the system to crash, resulting in partial updating.
+
+As a database, atomicity is a must-have feature. A single DML statement can be regarded as a minimal transaction. In terms of transactions, atomicity guarantees that each transaction is treated as a single "unit", which either succeeds completely or fails completely. A guarantee of atomicity prevents updates to the database from occurring only partially, which can cause greater problems than rejecting the whole series outright.
+
+## Design
+
+![](images/0059-dml-atomicity/redo_commit.svg)
+
+To provide atomicity for DML, we can use the concept of redo log or WAL (write-ahead logging). This RFC proposes a redo log and a commit log as internal states for `DMLExecutor`.
+
+Any insert, delete, or update executor should follow this contract:
+
+- Send `Begin(txn_id)` to `DMLExecutor`, where `txn_id` is a transaction ID that can be generated using the same algorithm as `row_id`.
+- Send `Chunk(txn_id)` to `DMLExecutor`.
+- Send `End(txn_id)` to `DMLExecutor`.
+
+(Note that all of the above messages must be sent to the same `DMLExecutor` e.g. using `hash(txn_id)` .)
+
+
+As for DMLExecutor, maintain two internal states: 
+1. create table redo_log(txn_id, ops, column_1, …, column_N). ops and columns are exactly the same meanings as the stream chunk.
+2. create table commit_log(txn_id, state). The value of the state column could be uncommitted or committed
```

Comment:
should we delete all the uncommitted txn because we do not know if there are some records lost?

Reply:
Yes. According to the contract,  after recovery,  if it is found that the state of the commit log is `uncommitted`, it means this transaction hasn't passed the commit point, so we have to delete all the records.